### PR TITLE
create client for substrate chains 

### PR DESCRIPTION
**Pull request by oshorefueled**
This PR implements the required provider and query methods to create light clients across substrate and cosmos chains.

### CODE REVIEW

File: dev-env-sub [hunk 1]
```diff
@@ -0,0 +1,62 @@
+#!/bin/bash -e
```
Comment (jtieri) [inline dev-env-sub]: Can we move this file into the directory with the other scripts/testing stuff? `./relayer/examples/demo/`

File: relayer/chains/substrate/key_test.go [hunk 2]
```diff
@@ -14,10 +15,11 @@ func TestKeyRestoreAndRetrieve(t *testing.T) {
 	mnemonic := "blind master acoustic speak victory lend kiss grab glad help demand hood roast zone lend sponsor level cheap truck kingdom apology token hover reunion"
 	expectedAddress := "5Hn67YZ75F3XrHiJAtiscJMGQ4zFNw9e45CNfLuxL6vEVYz8"
 
-	provider, err := getTestProvider()
-	require.Nil(t, err)
+	var provider = substrate.SubstrateProvider{}
```
Comment (jtieri) [inline relayer/chains/substrate/key_test.go]: `provider := substrate.SubstrateProvider{}` is slightly less verbose and achieves the same thing

File: relayer/chains/substrate/key_test.go [hunk 3]
```diff
@@ -14,10 +15,11 @@ func TestKeyRestoreAndRetrieve(t *testing.T) {
...
+	var err error
```
Comment (jtieri) [inline relayer/chains/substrate/key_test.go]: It looks like this variable declaration could be removed

File: relayer/chains/substrate/query.go [hunk 4]
```diff
@@ -479,6 +491,57 @@ func (sp *SubstrateProvider) QueryHeaderAtHeight(ctx context.Context, height int
...
+		return nil, fmt.Errorf("queried height is not finalized")
+	}
+
+	return sp.QueryLatestIBCHeader()
+}
+
+// QueryLatestIBCHeader returns the IBCHeader at the latest finalized relay chain height. It uses the relay chain height
+// which is set in the SubstrateProvider when the QueryLatestHeight method is called.
+func (sp *SubstrateProvider) QueryLatestIBCHeader() (provider.IBCHeader, error) {
+	var relayChainHeight uint64
```
Comment (jtieri) [inline relayer/chains/substrate/query.go]: I don't see a reason why this variable needs to be declared before initialization down on L521

File: relayer/chains/substrate/query.go [hunk 5]
```diff
@@ -479,6 +491,57 @@ func (sp *SubstrateProvider) QueryHeaderAtHeight(ctx context.Context, height int
...
+		return nil, err
+	}
+
+	header, err := sp.constructBeefyHeader(blockHash, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// reset latest queried relay height after IBC Header is queried
+	sp.LatestQueriedRelayChainHeight = 0
```
Comment (jtieri) [inline relayer/chains/substrate/query.go:532]: Why does this need to be reset after querying for an IBC header?

File: relayer/provider/matcher.go [hunk 6]
```diff
@@ -77,9 +81,11 @@ func tendermintMatcher(ctx context.Context, src, dst ChainProvider, existingClie
 			return "", tmclient.ErrTrustingPeriodExpired
 		}
 
+		fmt.Printf("\nquery ibc header %v \n \n", existingClientState.GetLatestHeight().GetRevisionHeight())
```
Comment (jtieri) [inline relayer/provider/matcher.go]: Looks like some debug output that may need to be removed.

File: relayer/provider/matcher.go [hunk 7]
```diff
@@ -77,9 +81,11 @@ func tendermintMatcher(ctx context.Context, src, dst ChainProvider, existingClie
...
 		// Construct a header for the consensus state of the counterparty chain.
 		ibcHeader, err := dst.QueryIBCHeader(ctx, int64(existingClientState.GetLatestHeight().GetRevisionHeight()))
 		if err != nil {
+			fmt.Printf("error quering header at height %v \n \n", existingClientState.GetLatestHeight().GetRevisionHeight())
```
Comment (jtieri) [inline relayer/provider/matcher.go]: Looks like some debug output that may need to be removed.

File: relayer/provider/matcher.go [hunk 8]
```diff
@@ -118,3 +124,69 @@ func isMatchingTendermintClient(a, b tmclient.ClientState) bool {
...
+		}
+
+		exportedConsState, err := clienttypes.UnpackConsensusState(consensusStateResp.ConsensusState)
+		if err != nil {
+			return "", err
+		}
+
+		existingConsensusState, ok := exportedConsState.(*beefyclienttypes.ConsensusState)
+		if !ok {
+			return "", fmt.Errorf("got type(%T) expected type(*tmclient.ConsensusState)", exportedConsState)
```
Comment (jtieri) [inline relayer/provider/matcher.go]: This should be `got type(%T) expected type(*beefyclienttypes.ConsensusState)`

File: relayer/provider/matcher.go [hunk 9]
```diff
@@ -118,3 +124,69 @@ func isMatchingTendermintClient(a, b tmclient.ClientState) bool {
...
+
+		// Construct a header for the consensus state of the counterparty chain.
+		ibcHeader, err := dst.QueryIBCHeader(ctx, int64(existingClientState.GetLatestHeight().GetRevisionHeight()))
+		if err != nil {
+			return "", err
+		}
+
+		consensusState, ok := ibcHeader.ConsensusState().(*beefyclienttypes.ConsensusState)
+		if !ok {
+			return "", fmt.Errorf("got type(%T) expected type(*tmclient.ConsensusState)", consensusState)
```
Comment (jtieri) [inline relayer/provider/matcher.go]: This should be `got type(%T) expected type(*beefyclienttypes.ConsensusState)`

**Review by jtieri — CHANGES_REQUESTED** [review]
Just a few nits but otherwise this looks good

File: relayer/chains/substrate/query.go (hunk 5 above)
Comment (oshorefueled) [reply relayer/chains/substrate/query.go:532]: The reason for resetting the variable is to make sure that QueryLatestHeight is called before calling QueryIBCHeader. When QueryHeaderAtHeight is called in the relayer, the returned height from QueryLatestHeight is passed to the method as the height argument. In the substrate provider, calling QueryLatestHeight returns the latest finalized parachain height. However, to fetch the header, we need to query the relaychain, only problem is there's no way to fetch the header from the relaychain using the para height. So, we cache the latest finalized relaychain height when QueryLatestHeight is called and reset it when the QueryHeaderAtHeight is called. That way, we make sure we're using the latest finalized relaychain height when QueryHeaderAtHeight is called.

**Review by jtieri — APPROVED** [review]

**Pull request merged**